File: card-infra/src/main/java/com/jean202/cardmizer/infra/persistence/jpa/JpaSpendingRecordAdapter.java
package com.jean202.cardmizer.infra.persistence.jpa;

import com.jean202.cardmizer.common.Money;
import com.jean202.cardmizer.core.application.ResourceNotFoundException;
import com.jean202.cardmizer.core.domain.CardId;
import com.jean202.cardmizer.core.domain.SpendingPeriod;
import com.jean202.cardmizer.core.domain.SpendingRecord;
import com.jean202.cardmizer.core.port.out.DeleteSpendingRecordPort;
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsByCardAndPeriodPort;
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsPort;
import com.jean202.cardmizer.core.port.out.SaveSpendingRecordPort;
import java.time.LocalDate;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@Primary
@Transactional
public class JpaSpendingRecordAdapter implements
        LoadSpendingRecordsPort,
        SaveSpendingRecordPort,
        LoadSpendingRecordsByCardAndPeriodPort,
        DeleteSpendingRecordPort {

    private final JpaSpendingRecordRepository repository;

    public JpaSpendingRecordAdapter(JpaSpendingRecordRepository repository) {
        this.repository = repository;
    }

    @Override
    @Transactional(readOnly = true)
    public List<SpendingRecord> loadByPeriod(SpendingPeriod period) {
        LocalDate startDate = period.yearMonth().atDay(1);
        LocalDate endDate = period.yearMonth().atEndOfMonth();
        return repository.findByDeletedFalseAndSpentOnBetweenOrderBySpentOnAscIdAsc(startDate, endDate).stream()
                .map(this::toDomain)
                .toList();
    }

    @Override
    @Transactional(readOnly = true)
    public List<SpendingRecord> loadByPeriodAndCard(SpendingPeriod period, CardId cardId) {
        LocalDate startDate = period.yearMonth().atDay(1);
        LocalDate endDate = period.yearMonth().atEndOfMonth();
        if (cardId != null) {
            return repository.findByDeletedFalseAndCardIdAndSpentOnBetweenOrderBySpentOnDescIdDesc(
                    cardId.value(), startDate, endDate).stream()
                    .map(this::toDomain)
                    .toList();
        }
        return repository.findByDeletedFalseAndSpentOnBetweenOrderBySpentOnDescIdDesc(startDate, endDate).stream()
                .map(this::toDomain)
                .toList();
    }

    @Override
    public void save(SpendingRecord spendingRecord) {
        repository.save(toEntity(spendingRecord));
    }

    @Override
    public void delete(UUID id) {
        JpaSpendingRecordEntity entity = repository.findById(id)
                .filter(e -> !e.isDeleted())
                .orElseThrow(() -> new ResourceNotFoundException("Spending record not found: " + id));
        entity.setDeleted(true);
        repository.save(entity);
    }

    private SpendingRecord toDomain(JpaSpendingRecordEntity entity) {
        return new SpendingRecord(
                entity.getId(),
                new CardId(entity.getCardId()),
                Money.won(entity.getAmount()),
                entity.getSpentOn(),
                entity.getMerchantName(),
                entity.getMerchantCategory(),
                parseTags(entity.getPaymentTags())
        );
    }

    private JpaSpendingRecordEntity toEntity(SpendingRecord spendingRecord) {
        return new JpaSpendingRecordEntity(
                spendingRecord.id(),
                spendingRecord.cardId().value(),
                spendingRecord.amount().amount(),
                spendingRecord.spentOn(),
                spendingRecord.merchantName(),
                spendingRecord.merchantCategory(),
                formatTags(spendingRecord.paymentTags()),
                false
        );
    }

    private String formatTags(Set<String> paymentTags) {
        if (paymentTags == null || paymentTags.isEmpty()) {
            return "";
        }
        return paymentTags.stream().sorted().collect(Collectors.joining(","));
    }

    private Set<String> parseTags(String serializedPaymentTags) {
        if (serializedPaymentTags == null || serializedPaymentTags.isBlank()) {
            return Set.of();
        }

        Set<String> tags = new LinkedHashSet<>();
        for (String rawTag : serializedPaymentTags.split(",")) {
            if (!rawTag.isBlank()) {
                tags.add(rawTag);
            }
        }
        return Set.copyOf(tags);
    }
}

File: card-infra/src/main/kotlin/com/jean202/cardmizer/infra/persistence/jpa/JpaSpendingRecordAdapter.kt
package com.jean202.cardmizer.infra.persistence.jpa

import com.jean202.cardmizer.common.Money
import com.jean202.cardmizer.core.application.ResourceNotFoundException
import com.jean202.cardmizer.core.domain.CardId
import com.jean202.cardmizer.core.domain.SpendingPeriod
import com.jean202.cardmizer.core.domain.SpendingRecord
import com.jean202.cardmizer.core.port.out.DeleteSpendingRecordPort
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsByCardAndPeriodPort
import com.jean202.cardmizer.core.port.out.LoadSpendingRecordsPort
import com.jean202.cardmizer.core.port.out.SaveSpendingRecordPort
import org.springframework.context.annotation.Primary
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Component
@Primary
@Transactional
class JpaSpendingRecordAdapter(
    private val repository: JpaSpendingRecordRepository,
) : LoadSpendingRecordsPort, SaveSpendingRecordPort, LoadSpendingRecordsByCardAndPeriodPort, DeleteSpendingRecordPort {

    @Transactional(readOnly = true)
    override fun loadByPeriod(period: SpendingPeriod): List<SpendingRecord> {
        val startDate = period.yearMonth.atDay(1)
        val endDate = period.yearMonth.atEndOfMonth()
        return repository.findByDeletedFalseAndSpentOnBetweenOrderBySpentOnAscIdAsc(startDate, endDate)
            .map { toDomain(it) }
    }

    @Transactional(readOnly = true)
    override fun loadByPeriodAndCard(period: SpendingPeriod, cardId: CardId?): List<SpendingRecord> {
        val startDate = period.yearMonth.atDay(1)
        val endDate = period.yearMonth.atEndOfMonth()
        return if (cardId != null) {
            repository.findByDeletedFalseAndCardIdAndSpentOnBetweenOrderBySpentOnDescIdDesc(
                cardId.value, startDate, endDate,
            )
        } else {
            repository.findByDeletedFalseAndSpentOnBetweenOrderBySpentOnDescIdDesc(startDate, endDate)
        }.map { toDomain(it) }
    }

    override fun save(spendingRecord: SpendingRecord) {
        repository.save(toEntity(spendingRecord))
    }

    override fun delete(id: UUID) {
        val entity = repository.findById(id)
            .filter { !it.deleted }
            .orElseThrow { ResourceNotFoundException("Spending record not found: $id") }
        entity.deleted = true
        repository.save(entity)
    }

    private fun toDomain(entity: JpaSpendingRecordEntity) = SpendingRecord(
        id = entity.id,
        cardId = CardId(entity.cardId),
        amount = Money.won(entity.amount),
        spentOn = entity.spentOn,
        merchantName = entity.merchantName,
        merchantCategory = entity.merchantCategory,
        paymentTags = parseTags(entity.paymentTags),
    )

    private fun toEntity(record: SpendingRecord) = JpaSpendingRecordEntity(
        id = record.id,
        cardId = record.cardId.value,
        amount = record.amount.amount,
        spentOn = record.spentOn,
        merchantName = record.merchantName,
        merchantCategory = record.merchantCategory,
        paymentTags = formatTags(record.paymentTags),
        deleted = false,
    )

    private fun formatTags(paymentTags: Set<String>): String =
        paymentTags.sorted().joinToString(",")

    private fun parseTags(serialized: String): Set<String> =
        if (serialized.isBlank()) emptySet()
        else serialized.split(",").filter { it.isNotBlank() }.toSet()
}
